# Assignment: Data Wrangling
### `! git clone https://github.com/ds4e/wrangling`
### Do Q1 and Q2, and either Q3 or Q4, for a total of 3 questions.

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)


In [1]:
# importing stuff
import numpy as np 
import pandas as pd 

In [2]:
# fetching data and printing out to look at it
df = pd.read_csv('./hw data/airbnb_hw.csv')
df.head(20)

# make every price a number and not a string
df['Price'] = pd.to_numeric(df['Price'], errors='coerce') 

# create a missing dummy, 1 = is missing & 0 = non-null
df['Price_nan'] = df['Price'].isnull() 

# describe the Price column
print("After cleaning: \n" , df['Price'].describe())

# print out number of missing values by using the Price_nan column
print("Total missing values: " , sum(df['Price_nan']))

After cleaning: 
 count    30297.000000
mean       153.357032
std        108.895698
min         10.000000
25%         80.000000
50%        125.000000
75%        192.000000
max        999.000000
Name: Price, dtype: float64
Total missing values:  181


2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 


In [3]:
# fetching and printing the data
da = pd.read_csv('./hw data/mn_police_use_of_force.csv')
da.head(20)

# print out what are the possible values for the column
print(da['subject_injury'].unique())

# replacing the NaN with np.nan
da['subject_injury'] = da['subject_injury'].replace('NaN', np.nan)

# print out number of values for each option, use dropna=False to include NaN in the counts
print("options: \n" , da['subject_injury'].value_counts(dropna=False))

# number of NaN values 
nan_count = da['subject_injury'].isna().sum()
# total number of values
total = da.shape[0] 
# proportion of missing values 
prop = nan_count / total

# print proportion 
print("Proportion of missing values:", prop)


[nan 'No' 'Yes']
options: 
 subject_injury
NaN    9848
Yes    1631
No     1446
Name: count, dtype: int64
Proportion of missing values: 0.7619342359767892


In [4]:
# cross-tabulating subject_injury and force_type 
c = pd.crosstab(da['force_type'], da['subject_injury'], dropna=False)
print (c)

subject_injury                 No   Yes   NaN
force_type                                   
Baton                           0     2     2
Bodily Force                 1093  1286  7051
Chemical Irritant             131    41  1421
Firearm                         2     0     0
Gun Point Display              33    44    27
Improvised Weapon              34    40    74
Less Lethal                     0     0    87
Less Lethal Projectile          1     2     0
Maximal Restraint Technique     0     0   170
Police K9 Bite                  2    44    31
Taser                         150   172   985


The proportion of values for subject_injury that are missing is 0.762. Yes, this is a concern. A proportion of 0.762 means that 76% of the values are missing, which is the majority of the values for this category. Considering that if the subject was injured or not is kind of an important category, it is very concerning that this column is missing the majority of its data values. 

The cross-tabulate results shows the options for subject injury as the columns and options for force type as the rows. We can see that for 6 out of the 11 types of force have more missing values than either a "yes" or "no" for subject injury when that force type was used. This is also concerning because that is more than half of the force types. 

Ones that are especially notable are for Bodily Force, Chemical Irritant, Less Lethal, Maximal Restraint Technique, and Taser. These are the force types that have the most discrepancies between the "yes" or "no" and NaN for subject injury. It could be safe to assume that the reason for the high number of missing values for these force types is because these are some of the stronger and more violent force types, and the police officers don't want to report that there was subject injury when they used more force than might have been necessary. 

3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.

In [5]:
# fetching and printing data
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
db = pd.read_csv(url,low_memory=False)
db.head(20)

InternalStudyID REQ_REC# Defendant_Sex Defendant_Race Defendant_BirthYear  \
0         ADI00001        1             M              W                1986   
1         ADI00007        3             M              B                1956   
2         ADI00008        4             M              W                1990   
3         CDI00036        6             M              B                1989   
4         CDI00038        7             F              W                1988   
5         CDI00039        8             M              B                1978   
6         CDI00041        9             F              B                1981   
7         CDI00043       10             M              B                1999   
8         CDI00044       11             M              B                1990   
9         CDI00046       12             M              W                1996   
10        CDI00048       13             M              W                1951   
11        CDI00050       14             M              B                1992   
12        CDI00055       15             F              W                1982   
13        CDI00062       16             F              W                1982   
14        CDI00088       17             M              W                1986   
15        CDI00089       18             F              W                1979   
16        CDI00090       19             M              W                1994   
17        CDI00091       20             M              W                1976   
18        CDI00092       21             F              W                1978   
19        CDI00093       22             M              W                1968   

   Defendant_Age  Defendant_AgeGroup Defendant_AgeatCurrentArrest  \
0             31                   3                           31   
1             60                   6                           60   
2             27                   3                           27   
3             27                   3                           27   
4             28                   3                           28   
5             39                   4                           39   
6             35                   3                           35   
7             18                   2                           18   
8             27                   3                           27   
9             21                   2                           21   
10            66                   7                           66   
11            25                   2                           25   
12            35                   3                           35   
13            35                   3                           35   
14            31                   3                           31   
15            38                   4                           38   
16            22                   2                           22   
17            40                   4                           40   
18            39                   4                           39   
19            48                   5                           48   

    Defendant_AttorneyTypeAtCaseClosure  Defendant_IndigencyStatus  ...  \
0                                    99                         99  ...   
1                                     9                          9  ...   
2                                     9                          9  ...   
3                                     0                          0  ...   
4                                     0                          0  ...   
5                                     0                          0  ...   
6                                     1                          1  ...   
7                                     0                          0  ...   
8                                     1                          1  ...   
9                                     1                          1  ...   
10                                    0        

In [6]:
print(db['WhetherDefendantWasReleasedPretrial'].unique())
print(db['WhetherDefendantWasReleasedPretrial'])
print(db['WhetherDefendantWasReleasedPretrial'].value_counts(dropna=False))

[9 0 1]
0        9
1        0
2        0
3        0
4        1
        ..
22981    1
22982    1
22983    1
22984    1
22985    1
Name: WhetherDefendantWasReleasedPretrial, Length: 22986, dtype: int64
WhetherDefendantWasReleasedPretrial
1    19154
0     3801
9       31
Name: count, dtype: int64


4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.

In [7]:
# fetching data
# pip install xlrd
ds = pd.read_excel('./hw data/shark_attacks.xls')
ds.head(20)

Date    Year        Type        Country  \
0   2025-01-23 00:00:00  2025.0  Unprovoked      Australia   
1   2025-01-22 00:00:00  2025.0  Unprovoked      Australia   
2   2025-01-11 00:00:00  2025.0    Provoked            USA   
3   2025-01-02 00:00:00  2025.0  Unprovoked  New Caledonia   
4   2025-01-02 00:00:00  2025.0  Unprovoked      Australia   
5   2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
6   2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
7   2024-12-28 00:00:00  2024.0  Unprovoked      Australia   
8   2024-12-02 00:00:00  2024.0  Unprovoked      Australia   
9   2024-12-01 00:00:00  2024.0  Unprovoked      Australia   
10  2024-11-30 00:00:00  2024.0  Unprovoked       Thailand   
11  2024-11-19 00:00:00  2024.0  Unprovoked    New Zealand   
12  2024-11-15 00:00:00  2024.0  Unprovoked       Maldives   
13  2024-11-01 00:00:00  2024.0  Unprovoked         Hawaii   
14  2024-10-31 00:00:00  2024.0  Unprovoked      Australia   
15  2024-10-25 00:00:00  2024.0  Unprovoked            USA   
16  2024-10-14 00:00:00  2024.0  Unprovoked       Honduras   
17  2024-10-11 00:00:00  2024.0  Unprovoked            USA   
18  2024-09-17 00:00:00  2024.0  Unprovoked       Honduras   
19  2024-09-16 00:00:00  2024.0  Unprovoked        Morocco   

                  State                                           Location  \
0     Western Australia                                     Shipwreck Cove   
1     Western Australia                                     Cheynes beach    
2                Hawaii                      Off Haleiwa Boat Harbour Oahu   
3          Grande Terre                        Islet of Kendek near Koumac   
4       South Australia            Granites Beach near Westall Streaky Bay   
5   North of Marsa Alam                                            Red Sea   
6   North of Marsa Alam                                            Red Sea   
7            Queensland                   Humpy Island Great Keppel Island   
8            Queensland     Curtis Island Near Gladstone QLD central coast   
9     Western Australia  Greenough River Mouth Beach Cape Burney 12km s...   
10   Phang Nga Province                                     Khao Lak Beach   
11      Bay of Waitangi                                    Chatham Islands   
12                Kaafu                                           Maafushi   
13                 Maui                       Sand Piles Waiehu Beach Park   
14           Queensland                                 Whitsunday Islands   
15              Florida                       Bathtub Beach, Martin County   
16                 Tela                                          Atlantida   
17              Florida                             Brevard County Orlando   
18       Trujillo Colon                                    Puerto Castillo   
19     Southern Morocco                                     West of Dakhla   

                       Activity                                    Name Sex  \
0                       Bathing                                    Boy    M   
1                       Surfing                             Dale Kittow   M   
2                        Diving  Male not stated was a dive tour worker   M   
3                  Spearfishing                        Robert Cuewapuru   M   
4                       Surfing                           Lance Appleby   M   
5                  SCUBA Diving                         Peppino Fappani   M   
6                  SCUBA Diving                       Gianluca Di Gioia   M   
7                  Spearfishing                            Luke Walford   M   
8                  Spearfishing                              Not stated   M   
9                       Surfing                  James Sweetman (child)   M   
10                     Swimming                              Elke Maier   F   
11                       Diving                     Jade Kahukore-Dixon   M   
12                       Diving                   

2. Drop any columns that do not contain data.

In [8]:
# making for loop that prints out the possible values for each column to determine which
# columns don't contain data 
# 23 columns
# print(ds.columns)
# empty = ds.isnull().all()
# print(empty.tolist())
# print(ds['pdf'].unique())
ds_clean = ds.dropna(axis = 1, how='all')
ds_clean.head(20)

Date    Year        Type        Country  \
0   2025-01-23 00:00:00  2025.0  Unprovoked      Australia   
1   2025-01-22 00:00:00  2025.0  Unprovoked      Australia   
2   2025-01-11 00:00:00  2025.0    Provoked            USA   
3   2025-01-02 00:00:00  2025.0  Unprovoked  New Caledonia   
4   2025-01-02 00:00:00  2025.0  Unprovoked      Australia   
5   2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
6   2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
7   2024-12-28 00:00:00  2024.0  Unprovoked      Australia   
8   2024-12-02 00:00:00  2024.0  Unprovoked      Australia   
9   2024-12-01 00:00:00  2024.0  Unprovoked      Australia   
10  2024-11-30 00:00:00  2024.0  Unprovoked       Thailand   
11  2024-11-19 00:00:00  2024.0  Unprovoked    New Zealand   
12  2024-11-15 00:00:00  2024.0  Unprovoked       Maldives   
13  2024-11-01 00:00:00  2024.0  Unprovoked         Hawaii   
14  2024-10-31 00:00:00  2024.0  Unprovoked      Australia   
15  2024-10-25 00:00:00  2024.0  Unprovoked            USA   
16  2024-10-14 00:00:00  2024.0  Unprovoked       Honduras   
17  2024-10-11 00:00:00  2024.0  Unprovoked            USA   
18  2024-09-17 00:00:00  2024.0  Unprovoked       Honduras   
19  2024-09-16 00:00:00  2024.0  Unprovoked        Morocco   

                  State                                           Location  \
0     Western Australia                                     Shipwreck Cove   
1     Western Australia                                     Cheynes beach    
2                Hawaii                      Off Haleiwa Boat Harbour Oahu   
3          Grande Terre                        Islet of Kendek near Koumac   
4       South Australia            Granites Beach near Westall Streaky Bay   
5   North of Marsa Alam                                            Red Sea   
6   North of Marsa Alam                                            Red Sea   
7            Queensland                   Humpy Island Great Keppel Island   
8            Queensland     Curtis Island Near Gladstone QLD central coast   
9     Western Australia  Greenough River Mouth Beach Cape Burney 12km s...   
10   Phang Nga Province                                     Khao Lak Beach   
11      Bay of Waitangi                                    Chatham Islands   
12                Kaafu                                           Maafushi   
13                 Maui                       Sand Piles Waiehu Beach Park   
14           Queensland                                 Whitsunday Islands   
15              Florida                       Bathtub Beach, Martin County   
16                 Tela                                          Atlantida   
17              Florida                             Brevard County Orlando   
18       Trujillo Colon                                    Puerto Castillo   
19     Southern Morocco                                     West of Dakhla   

                       Activity                                    Name Sex  \
0                       Bathing                                    Boy    M   
1                       Surfing                             Dale Kittow   M   
2                        Diving  Male not stated was a dive tour worker   M   
3                  Spearfishing                        Robert Cuewapuru   M   
4                       Surfing                           Lance Appleby   M   
5                  SCUBA Diving                         Peppino Fappani   M   
6                  SCUBA Diving                       Gianluca Di Gioia   M   
7                  Spearfishing                            Luke Walford   M   
8                  Spearfishing                              Not stated   M   
9                       Surfing                  James Sweetman (child)   M   
10                     Swimming                              Elke Maier   F   
11                       Diving                     Jade Kahukore-Dixon   M   
12                       Diving                   

3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?

4. Clean the Age variable and make a histogram of the ages of the victims.

5. What proportion of victims are male?

6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.

8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?

9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

**Q3.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q4.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?

The US Census recently gathered data on race in the 2020 Census. They had specific checkboxes for major racial categories, such as White, Black or African American, American Indian or Alaska Native, Asian, with subcategories, Native Hawaiian or other Pacific Islander, and Other, with a write-in option. 

2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?

We gather these data for legal, social, economic, and policy-making purposes. These purposes include ensuring fair representation, resources are allocated equitably, and helping to understand demographic trends. In term of politics, having race data helps policy-makers identify racial disparities in many aspects of society, such as healthcare, housing, and economic opportunities. In terms of society, having race data helps tailor public health programs to the area so that it is the most effective and beneficial to the residents. Data quality matters because not having accurate representation can distort political power, funding, and resource distribution. 

3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?


4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.


5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?


6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?